Rework unloadability fix

janvorli opened the pull request:
This change reworks the recent fix of dependency between two collectible `AssemblyLoadContext`'s when one context uses the other one for resolving its assemblies. There was a problem with the previous solution that has shown up in the newly added test when running with GCStress 3 enabled. 
When we create the reference between the related native LoaderAllocator instances, the managed `LoaderAllocator` of the `AssemblyLoadContext` that was used to resolve the assembly needs to be alive so that we can create a handle to it in the context of the parent `AssemblyLoadContext` and keep it alive for the lifetime of the parent `AssemblyLoadContext`. 
With the previous change, if the `AssemblyLoadContext` that was used to resolve the assembly was collected before the `AssemblyLoadContext.Load` override or the related Resolving event returned and the managed representation of the resolved assembly was collected too, we would not be able to create managed `System.Reflection.Type` instances for the types from the resolved assembly after that. So e.g. calling `GetInterfaces` on a `Type` would fail or crash.

This change fixes the problem by moving the creation of reference between the native `LoaderAllocator`s to the `RuntimeInvokeHostAssemblyResolver` where the managed `Assembly` reference is still alive and thus the related managed `LoaderAllocator` must be alive too. 

Close #68112

janvorli committed:
Rework the unloadability fix

The recent fix had a problem when GC collected the managed Assembly
loaded via the Load override or the Resolving event before we added the
reference between the related native LoaderAllocators. The managed
LoaderAllocator of the ALC we've loaded the Assembly into was collected
and we couldn't create the managed Type objects for the interfaces from
that assembly in the GetInterfaces call on a type that implemented
those.

This change fixes it by moving the creation of reference between the
native LoaderAllocators to the RuntimeInvokeHostAssemblyResolver where
we still have a live managed reference to the loaded Assembly.

janvorli committed:
Use LoaderAllocator from binder only

janvorli committed:
Pass binder instead of LoaderAllocator

janvorli committed:
Move clearing of the m_pAssemblyLoaderAllocator to a later stage

elinor-fung commented on src/coreclr/binder/assemblybindercommon.cpp:
```diff
@@ -1153,9 +1154,10 @@ namespace BINDER_SPACE
 
 #if !defined(DACCESS_COMPILE)
 HRESULT AssemblyBinderCommon::BindUsingHostAssemblyResolver(/* in */ INT_PTR pManagedAssemblyLoadContextToBindWithin,
```
We can probably do some cleanup around just getting the managed ALC pointer from the AssemblyBinder now that we have the binder - after this change goes in, since we want to backport this.

janvorli commented on src/coreclr/vm/appdomain.cpp:
```diff
@@ -5115,6 +5084,17 @@ HRESULT RuntimeInvokeHostAssemblyResolver(INT_PTR pManagedAssemblyLoadContextToB
 
+            // For collectible assemblies, ensure that the parent loader allocator keeps the assembly's loader allocator
+            // alive for all its lifetime.
+            if (pDomainAssembly->IsCollectible())
```
Hmm, but we need to create the link even in that case (and also in case of non-default non-collectible context) or rather prevent that from happening by throwing an exception, otherwise the ALC we've resolved from would get collected and we would end up with the issue as before. I have thought we do prevent that already, but local testing has revealed we don't.

elinor-fung replied:
Oh, hm, yeah, we are definitely missing that case then, since `EnsureReference` also won't add the reference for non-collectible.

Looks like we error out on collectible assemblies resolved by `AppDomain.AssemblyResolve`.
https://github.com/dotnet/runtime/blob/eb9dd67ce9fc64c468cf5c795fda7057310c22fc/src/coreclr/vm/appdomain.cpp#L4640-L4643
But we don't check the results of other extension points.

janvorli replied:
@elinor-fung I've added the exception throwing for NULL parent loader allocator.

janvorli committed:
Throw exception on collectible assembly in non-collectible ALC

elinor-fung commented on src/coreclr/vm/appdomain.cpp:
```diff
@@ -5115,6 +5084,22 @@ HRESULT RuntimeInvokeHostAssemblyResolver(INT_PTR pManagedAssemblyLoadContextToB
+                if (pParentLoaderAllocator == NULL)
+                {
+                    // The AssemblyLoadContext for which we are resolving the the Assembly is not collectible.
+                    COMPlusThrow(kNotSupportedException, W("NotSupported_CollectibleBoundNonCollectible")); 
```
We should add a test for this (fine with doing it later).

janvorli replied:
Makes sense. I can extend the test I've added previously  to have new test case(s) for that.

elinor-fung replied:
Actually, I think this is technically a breaking change? I agree it is the right thing, but I'm not sure what the appetite for backporting to 6.0 would be (all about it for 7.0 though).

janvorli replied:
For the 6.0, I'll let the shiproom decide whether to do that or just omit the check. You are right that for some apps, this might accidentally work without this check and now it would always throw an exception.

elinor-fung replied:
Sounds good to me.

janvorli replied:
Done

elinor-fung reviewed (approved):
We should also file a docs issue for how we're now explicitly blocking the non-collectible to collectible assembly: https://github.com/dotnet/docs/issues/new?template=dotnet-breaking-change.md
Or I think marking this with the 'breaking-change' label gives you all the instructions for that.

janvorli committed:
Update the ResolvedFromDifferentContext test

Make it test that we throw an exception when the parent ALC is not
collectible.
